services/slam/fake/slam.go

```diff
@@ -0,0 +1,171 @@
+type SLAM struct {
+	generic.Echo
+	Name      string
+	dataCount int
+	logger    golog.Logger
+}
+
+// GetMap returns either a vision.Object or image.Image based on request mimeType.
+func (slamSvc *SLAM) GetMap(ctx context.Context, name, mimeType string, cp *referenceframe.PoseInFrame,
+	include bool, extra map[string]interface{},
+) (string, image.Image, *vision.Object, error) {
+	slamSvc.incrementDataCount()
+
+	var err error
+	var img image.Image
+	var vObj *vision.Object
+
+	switch mimeType {
+	case rdkutils.MimeTypePCD:
+		path := filepath.Clean(artifact.MustPath(fmt.Sprintf(pcdTemplate, slamSvc.dataCount)))
+		slamSvc.logger.Debug("Reading " + path)
+
+		f, err := os.Open(path)
+		if err != nil {
+			return "", nil, nil, err
+		}
+		defer utils.UncheckedErrorFunc(f.Close)
+
+		pc, err := pointcloud.ReadPCD(f)
+		if err != nil {
+			return "", nil, nil, err
+		}
+
+		vObj, err = vision.NewObject(pc)
+		if err != nil {
+			return "", nil, nil, err
+		}
+
+	case rdkutils.MimeTypeJPEG:
+		path := filepath.Clean(artifact.MustPath(fmt.Sprintf(pngTemplate, slamSvc.dataCount)))
+		slamSvc.logger.Debug("Reading " + path)
+		img, err = rimage.NewImageFromFile(path)
+
+	default:
+		return "", nil, nil, errors.New("received invalid mimeType for GetMap call")
+	}
+
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	return mimeType, img, vObj, nil
+}
+
+// Position returns a PoseInFrame of the robot's current location according to SLAM.
+func (slamSvc *SLAM) Position(ctx context.Context, name string, extra map[string]interface{}) (*referenceframe.PoseInFrame, error) {
+	path := filepath.Clean(artifact.MustPath(fmt.Sprintf(positionTemplate, slamSvc.dataCount)))
+	slamSvc.logger.Debug("Reading " + path)
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	substrings := strings.Split(string(data), " | ")
+
+	rawXYZ, err := extract(strings.Split(substrings[0], " "))
+	if err != nil {
+		return nil, err
+	}
+	xyz := r3.Vector{X: rawXYZ[0], Y: rawXYZ[1], Z: rawXYZ[2]}
+
+	rawAA, err := extract(strings.Split(substrings[1], " "))
+	if err != nil {
+		return nil, err
+	}
+
+	axisAngle := spatialmath.NewR4AA().AxisAngles()
+	axisAngle.RX, axisAngle.RY, axisAngle.RZ, axisAngle.Theta = rawAA[0], rawAA[1], rawAA[2], rawAA[3]
+	pose := spatialmath.NewPose(xyz, axisAngle)
+	pInFrame := referenceframe.NewPoseInFrame(name, pose)
+
+	return pInFrame, nil
+}
+
+// GetInternalState returns the internal state of a slam algo. Curently the internal state of cartogropher.
+func (slamSvc *SLAM) GetInternalState(ctx context.Context, name string) ([]byte, error) {
+	path := filepath.Clean(artifact.MustPath(fmt.Sprintf(internalStateTemplate, slamSvc.dataCount)))
+	slamSvc.logger.Debug("Reading " + path)
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	return data, nil
+}
+
+func (slamSvc *SLAM) incrementDataCount() {
+	slamSvc.dataCount = ((slamSvc.dataCount + 1) % maxDataCount)
```

Comment:
Maybe add a comment that we only ever expect one browser open for fake slam, so it's okay this isn't thread-safe?

Reply:
Good call. Done